### ISSUE TITLE
A small IFC file is stuck in an infinite loop generating geometry

### ISSUE THREAD

**Issue by carmenfan** [3drepo/3drepobouncer#633]
### Description
A customer IFC file of ~6MB is timing out. Upon inspection, it seems to be stuck on an infinite loop whilst trying to generate geometry.

Link to the model: https://asitesol.sharepoint.com/:f:/s/3DRepoTeamFolder/EvYJY_cTSr1CjSS6grflBRcBtCXPMDaIg6hAuvnafjogfQ?e=LSPGu5

The log is stuck on this stage forever until times out:
![image](https://github.com/3drepo/3drepobouncer/assets/11945337/2569b9f1-31ff-4ba9-ad7d-f589643c2cf6)

My suspicion is that it is trying to generate too high level LOD on a some tiny curves, resulting in the same vertice is being generated and it never hits the loop terminating condition (have seen something similar before in IFC Open shell years ago). So tweaking the LOD configurations may be the easiest fix as we are looking into this anyway with #632

**Comment by sebjf** [comment]
Recreated this issue in my local bouncer, and found where to set the angular resolution and linear tolerance in the import settings. These are the same as used by ODA, which is nice for #632 (they both seem to use a BRep underneath).

Unfortunately changing the LOD doesn't help the issue.

Using IfcConvert from the 0.7.0 build from IfcOpenShell website also locks up indefinitely, so the problem is in the latest version of the library.

Building IfcOpenShell from source and running IfcConvert under the debugger shows it breaks in `RelocatePCurvesToNewUOrigin`. (ShapeUpgrade_UnifySameDomain.cxx line ~360)

What happens in this function is that it tries to evaluate a sequence of curves on a face (edges in a contour). At the start of the loop it tries to remove the last evaluated edge from the list of edges (`theVEmap`), which should eventually leave this list empty. This list is never emptied however, causing the loop to run infinitely.

The `RemoveEdgeFromMap` method works by checking the edge against each entry using the IsSame member, which seems to be a superclass method.

The debugging symbols lose where they are here so its not obvious why this call fails.

**Comment by carmenfan** [comment]
@sebjf could it be that the problem lies in OCCT?

**Comment by sebjf** [comment]
Hi @carmenfan the problem is definitely in OCCT.

The fault occurs in [Unify Same Domain](https://dev.opencascade.org/doc/overview/html/occt_user_guides__shape_healing.html#occt_shg_4_4_9) Build() stage, which merges coincident faces/edges that lie on coincident surfaces/curves.

Specifically, its in `RelocatePCurvesToNewUorigin()` function.

This method iterates through all the Edges in a Shape (which sit on a curved surface). For each, it gets the Curve formed by the projection of the Edge (`CurEdge`) on the Surface (`CurPCurve`) and adds this to a Map, while also storing the starting Parameter, and the Point on the Curve at this Parameter.

In a second nested loop it iterates through all Edges that start at the same Vertex as `CurEdge` (`anEdge`) using a map of the edges that start at each vertex in the Shape `theVEmap`. It performs a similar operation for these (creating a new Curve, and adding it to a Map).

When this second loop exits, it picks up the next Edge in the Shape and repeats.

In the second nested loop, `CurEdge` is added to a map, `theUsedEdges` and removed from `theVEmap`. The outermost loop terminates when it can't find an Edge that does not exist in `theUsedEdges`.

The problem is the termination criteria of the second nested loop. This loop only terminates when there are no more edges in `theVEmap` for the starting vertex of `CurEdge`.

In Concrete.ifc however, this method encounters a 'CurEdge' that never triggers this criteria, causing it to get stuck in the loop indefinitely.

It is not entirely obvious why this criteria is never reached, but it seems to be an interaction between two parts of this method:

1. The `RemoveEdgeFromMap()` call which should remove 'CurEdge' from 'theVEmap' and therefore make it smaller on each iteration, and..
2. A check for whether the start Point of `anEdge` matches (is within a certain distance of) that of `CurEdge`.

This second check occurs before 'CurEdge' is updated, meaning if it fails 'CurEdge' goes straight back to be removed from 'theVEmap' - if this operation does not decrement 'theVEmap' (which it doesn't) the loop gets stuck.

I believe the intent behind how this is written is that it if `anEdge` is not valid for some reason, it doesn't try to create a new Curve for it. The problem is that if calling `RemoveEdgeFromMap()` doesn't result in `theVEmap` being reduced in size for whatever reason, this check will prevent `CurEdge` from advancing.

What is not clear to me at the moment is whether it should be possible to call `RemoveEdgeFromMap()` and *not* actually remove anything, because if not, a simple fix is to just report a failure to remove the edge, and terminate the inner loop here.

Another possible fix is to gate the curve updates differently, and make sure that 'CurEdge' is always updated by this loop.

**Comment by carmenfan** [comment]
@sebjf would updating OCCT fixes this? (building it takes forever though 😆). Might worth checking if they've identified the issue else where and have since fixed it

**Comment by sebjf** [comment]
As I've found 😆 (I had to rebuild with optimisation off to investigate properly!) 

It is a known issue: https://tracker.dev.opencascade.org/view.php?id=32949

Reported last year but without working steps to reproduce. Someone posted another example a couple of months ago, but no activity since.

**Comment by sebjf** [comment]
A straightforward fix is to detect the exact condition that leads to the loop (CurEdge is not removed from the theVEmap and is not changed by the third nested loop).

This can be done by modifying `RemoveEdgeFromMap` to return how many elements have been removed, and storing the value of `CurEdge` before it enters the curve construction loop.

If it's the case that `RemoveEdgeFromMap` should never fail to remove an edge, this test can be made simpler, but not by much! (We just don't need to check CurEdge a second time.)

As it is this implementation is highly efficient as it doesn't involve checking any of the structures again, and works with Concrete.ifc

![image](https://github.com/3drepo/3drepobouncer/assets/5058369/b3827e06-272c-4a2c-8abe-2670b953f863)

## Dev Notes ##

OpenShell 0.0.7 will make a copy of all dependencies locally, as source, which is nice. So modifications of OCCT can be attempted in-place. The debugger will find the source of the dependencies on its own but they won't automatically rebuild.

The OCCT project will have a Visual Studio solution already created for it, e.g. in 

`D:\3drepo\bouncer\IfcOpenShell-0.7.0\_deps\occt_git\_build-vs2019-Win32\OCCT.sln`

This builds into 

`D:\3drepo\bouncer\IfcOpenShell-0.7.0\_deps\occt_git\_build-vs2019-Win32\win32\vc14\libi`

The IfcOpenShell project takes the libraries from

`D:\3drepo\bouncer\IfcOpenShell-0.7.0\_deps-vs2019-Win32-installed\opencascade-7.5.3\win32\lib`

So, when modifying OCCT, the modified libraries must be copied from `_deps\occt_git\_build-vs2019-Win32\win32\vc14\libi` into `_deps-vs2019-Win32-installed\opencascade-7.5.3\win32\lib` manually.

**Comment by sebjf** [comment]
Regarding bouncer compatibility,

1. Builds of bouncer with OCCT-7.5.3, but the existing OpenShell-0.6.0 binaries, fail with an access violation. 
2. (OCCT-7.5.3 is the version targeted by the IfcOpenShell-0.7.0 release, which is only a couple of minor versions behind master, so seems a suitable version to aim for.) 
3. Building bouncer with the IfcOpenShell-0.7.0 available in binary form from the website fails with API incompatibilities.
4. Building [IfcOpenShell-0.6.0](https://github.com/IfcOpenShell/IfcOpenShell/tree/v0.6.0) with our version of OCCT-7.5.3 fails with a linker error (nothing to do with our OCCT, its creating too many linker objects even with incremental linking off).
5. Building bouncer with our version of OpenShell-0.6.0 ... *WAITING ON A WORKING BUILD OF IFCOPENSHELL*

(It is not clear which version of OCCT IfcOpenShell-0.6.0 was originally built with in order to backport the fixes - the github readme just points to the download pages.)


Regarding the PR, the file containing the fix has undergone a number of changes since 7.5.3, so we will probably need to re-create the issue in master and make a PR based there, while duplicating the fix in our own fork of 7.5.3 until we upgrade at a later date.

Using a modified version of IfcOpenShell-0.7.0, we have extracted just the shape causing the issue, however this does not import into Draw.exe built from the same tag/branch as used by IfcOpenShell-0.7.0, because the importer doesn't know about file version 3, even though it wrote it.

Adding a snippet to the beginning of DRAWEXE allows us to recreate the issue inside OCCT-7.5.3 (separate to IfcOpenShell, and the customer's file) and demonstrate the fix works.

However, adding the same snippet in master fails. Our fix works, but then another part of the application fails with an access violation. This time in another area, because the code checks if an edge is null, and if so does something, but assumes that thing succeeds when it doesn't - basically the same type of bug that we just fixed! (At least this one is fairly straightforward to fix as well...)

```
#include <BinTools.hxx>
#include <TopoDS.hxx>
#include <TopoDS_Edge.hxx>
#include <ShapeUpgrade_UnifySameDomain.hxx>
{
  TopoDS_Shape s;
  BinTools::Read(s, "D:\\3drepo\\ISSUE_633\\shape.brep");
  ShapeUpgrade_UnifySameDomain usd(s);
  usd.Build();
}
```
